In [1]:
import pandas as pd 
df=pd.read_csv("./data_for_modling.csv")

In [3]:
df.head()

data  trip_creation_time  \
0  training          1537410936   
1  training          1537410936   
2  training          1537410936   
3  training          1537410936   
4  training          1537410936   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  od_start_time  ...  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)     1537413692  ...   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)     1537413692  ...   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)     1537413692  ...   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)     1537413692  ...   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)     1537413692  ...   

   od_start_time^2  od_start_time osrm_time  od_start_time osrm_distance  \
0     2.363641e+18             1.691155e+10                 1.839562e+10   
1     2.363641e+18             3.074827e+10                 3.339924e+10   
2     2.363641e+18             4.304758e+10                 5.002667e+10   
3     2.363641e+18             6.149655e+10                 7.004764e+10   
4     2.363641e+18             6.764620e+10                 8.335565e+10   

   od_start_time route_type_enc  osrm_time^2  osrm_time osrm_distance  \
0                           0.0        121.0                 131.6183   
1                           0.0        400.0                 434.4860   
2                           0.0        784.0                 911.1060   
3                           0.0       1600.0                1822.4800   
4                           0.0       1936.0                2385.5964   

   osrm_time route_type_enc  osrm_distance^2  osrm_distance route_type_enc  \
0                       0.0       143.168404                           0.0   
1                       0.0       471.945210                           0.0   
2                       0.0      1058.819060                           0.0   
3                       0.0      2075.895844                           0.0   
4                       0.0      2939.602368                           0.0   

   route_type_enc^2  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 47 columns]

In [5]:
df.isna().sum()

data                              0
trip_creation_time                0
route_schedule_uuid               0
route_type                        0
trip_uuid                         0
source_center                     0
source_name                       0
destination_center                0
destination_name                  0
od_start_time                     0
od_end_time                       0
start_scan_to_end_scan            0
is_cutoff                         0
cutoff_factor                     0
cutoff_timestamp                  0
actual_distance_to_destination    0
actual_time                       0
osrm_time                         0
osrm_distance                     0
factor                            0
segment_actual_time               0
segment_osrm_time                 0
segment_osrm_distance             0
segment_factor                    0
od_deffrent_time                  0
numebre_dayofweek                 0
route_type_enc                    0
od_end_time.1               

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  int64  
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144867 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144867 non-null  object 
 9   od_start_time                   144867 non-null  int64  
 10  od_end_time                     144867 non-null  int64  
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

In [9]:
df_for_modeling=df[['od_deffrent_time','od_start_time','od_end_time','trip_creation_time','osrm_time route_type_enc','od_start_time osrm_distance','od_end_time osrm_distance','osrm_distance^2','route_type_enc']]

In [11]:
from sklearn.model_selection import train_test_split
target=df['start_scan_to_end_scan']
x_train , x_test , y_train , y_test = train_test_split(df_for_modeling,target , test_size=0.2)

In [12]:
from sklearn.preprocessing import StandardScaler 
scale=StandardScaler()
x_train=scale.fit_transform(x_train)
x_test=scale.transform(x_test)

In [13]:
from sklearn.model_selection import GridSearchCV 
from sklearn.svm import SVR
# tuned_parameters = [
#     {"kernel": ["linear"], "C": [1, 10]},
#     {"kernel": ["poly"], "C":[1, 10]},
#     {"kernel": ["rbf"], "gamma": [1e-3, 1e-4], "C": [1, 10]}
# ]
# svr_model = SVR()
# # grid_search = RandomizedSearchCV(svr_model, tuned_parameters, n_iter=5, cv=5, n_jobs=-1)
# from sklearn.model_selection import RandomizedSearchCV
# grid_search = RandomizedSearchCV(svr_model, tuned_parameters, n_iter=5, cv=5, n_jobs=-1)
# grid_search.fit(x_train,y_train)
# print("Best parameters:", grid_search.best_params_)



In [17]:
from sklearn.svm import SVR
svr_model = SVR(kernel='rbf', C=1.0, epsilon=0.1)


In [23]:
x_train.size

1043037

In [25]:
svr_model.fit(x_train, y_train)

SVR()

In [29]:
y_pred_cvm=svr_model.predict(x_test)

In [33]:
from sklearn.metrics import mean_squared_error 
import numpy as np
import math 
print("Linear Regression Model RMSE is: ", math.sqrt(mean_squared_error(y_test, y_pred_cvm)))
print("Linear Regression Model Training Score: ",svr_model.score(x_train, y_train))
print("Linear Regression Model Testing Score: ",svr_model.score(x_test, y_test))

Linear Regression Model RMSE is:  107.72329075295015
Linear Regression Model Training Score:  0.9893310217487097
Linear Regression Model Testing Score:  0.9892610440703089


In [ ]:
# from sklearn.model_selection import GridSearchCV
# from sklearn.preprocessing import PolynomialFeatures
# from sklearn.linear_model import LinearRegression
# param_grid = {'degree': [2, 3, 4, 5]}

# poly = PolynomialFeatures()
# model = LinearRegression()

# grid_search = GridSearchCV(estimator=poly, param_grid=param_grid, cv=5, n_jobs=-1)

# grid_search.fit(X_train, y_train)

# print("Best degree:", grid_search.best_params_)


In [61]:

poly = PolynomialFeatures(degree=3)  # Degree of the polynomial
X_train_poly = poly.fit_transform(x_train)
model = LinearRegression()
model.fit(X_train_poly, y_train)
X_test_poly = poly.transform(x_test)
y_pred = model.predict(X_test_poly)

In [45]:
y_pred = model.predict(X_test_poly)

In [63]:
print("Linear Regression Model RMSE is: ", math.sqrt(mean_squared_error(y_test, y_pred)))
print("Linear Regression Model Training Score: ", model.score(X_train_poly, y_train))
print("Linear Regression Model Testing Score: ", model.score(X_test_poly, y_test))

Linear Regression Model RMSE is:  0.28844381627855337
Linear Regression Model Training Score:  0.9999999230404614
Linear Regression Model Testing Score:  0.9999999230045074


In [65]:

poly = PolynomialFeatures(degree=2) 
X_train_poly = poly.fit_transform(x_train)
model = LinearRegression()
model.fit(X_train_poly, y_train)
X_test_poly = poly.transform(x_test)
y_pred = model.predict(X_test_poly)
y_pred_2_degreess= model.predict(X_test_poly)
print("Linear Regression Model RMSE is: ", math.sqrt(mean_squared_error(y_test, y_pred_2_degreess)))
print("Linear Regression Model Training Score: ", model.score(X_train_poly, y_train))
print("Linear Regression Model Testing Score: ", model.score(X_test_poly, y_test))

Linear Regression Model RMSE is:  0.28954164457495407
Linear Regression Model Training Score:  0.9999999225069912
Linear Regression Model Testing Score:  0.9999999224172964


In [67]:
## for redge and lasso

In [69]:
from sklearn.model_selection import RepeatedKFold 
from sklearn.linear_model import LassoCV
cv=RepeatedKFold(n_splits=5,n_repeats=3 , random_state=1)

In [71]:
cv=RepeatedKFold(n_splits=5,n_repeats=3 , random_state=1)
lasso=LassoCV(alphas=np.arange(0.1,10,0.1) , cv =cv , tol=1)
lasso.fit(x_train,y_train)
lasso_y_pred=lasso.predict(x_test)

In [73]:
print("Linear Regression Model RMSE is: ", math.sqrt(mean_squared_error(y_test, lasso_y_pred)))
print("Linear Regression Model Training Score: ",lasso.score(x_train, y_train))
print("Linear Regression Model Testing Score: ",lasso.score(x_test, y_test))

Linear Regression Model RMSE is:  0.30677033883987775
Linear Regression Model Training Score:  0.9999999125371086
Linear Regression Model Testing Score:  0.9999999129097438


In [75]:
from sklearn.linear_model import RidgeCV
cv=RepeatedKFold(n_splits=5,n_repeats=3 , random_state=1)
ridege=RidgeCV(alphas=np.arange(0.1,10,0.1) , cv =cv , scoring="neg_mean_squared_error")

In [77]:
ridege.fit(x_train,y_train)
ridege_y_pred=ridege.predict(x_test)
print(ridege.alpha_)

0.1


In [79]:
print("Linear Regression Model RMSE is: ", math.sqrt(mean_squared_error(y_test, ridege_y_pred)))
print("Linear Regression Model Training Score: ",ridege.score(x_train, y_train))
print("Linear Regression Model Testing Score: ",ridege.score(x_test, y_test))

Linear Regression Model RMSE is:  0.29031569091099796
Linear Regression Model Training Score:  0.9999999220194327
Linear Regression Model Testing Score:  0.9999999220019303


In [81]:
print("SVR RMSE is: ", math.sqrt(mean_squared_error(y_test, y_pred_cvm)))


print("polunomial regression 2 degress is: ", math.sqrt(mean_squared_error(y_test, y_pred_2_degreess)))

print("polunomial regression 3 deg is: ", math.sqrt(mean_squared_error(y_test, y_pred)))

print("Linear Regression with redge RMSE is: ", math.sqrt(mean_squared_error(y_test, ridege_y_pred)))

print("Linear Regression with laso RMSE is: ", math.sqrt(mean_squared_error(y_test, lasso_y_pred)))
 

SVR RMSE is:  107.72329075295015
polunomial regression 2 degress is:  0.28954164457495407
polunomial regression 3 deg is:  0.28954164457495407
Linear Regression with redge RMSE is:  0.29031569091099796
Linear Regression with laso RMSE is:  0.30677033883987775


<table>
    <tr>
        <tdstrongel </td>
        <td> RMSE </td>
        <td> Notes </td>
    </tr>
    <tr>
        <td> SVR </td>
        <td> 107.72 </td>
        <td id="svr_note"> Poor performance with high RMSE. </td>
    </tr>
    <tr>
        <td> Polynomial Regression (Degree 2 & 3) </td>
        <td> 0.29 </td>
        <td id="poly_note"> Best performance with low RMSE. No difference between degrees 2 and 3. </td>
    </tr>
    <tr>
        <td> Linear Regression (Ridge) </td>
        <td> 0.29 </td>
        <td id="ridge_note"> Performs similarly to polynomial regression, good generalization. </td>
    </tr>
    <tr>
        <td> Linear Regression (Lasso) </td>
        <td> 0.31 </td>
        <td id="lasso_note"> Slightly worse than Ridge, but still good performance. </td>
    </tr>
</table>
